FWCore/Integration/test/TestInterProcessProd.cc

```diff
@@ -0,0 +1,286 @@
+  void streamBeginRun(edm::StreamID, edm::Run const&, edm::EventSetup const&) const final;
+  void streamEndRun(edm::StreamID, edm::Run const&, edm::EventSetup const&) const final;
+  void globalEndRun(edm::Run const&, edm::EventSetup const&) const final {}
+  void globalEndRunProduce(edm::Run&, edm::EventSetup const&) const final;
+
+  void globalBeginLuminosityBlockProduce(edm::LuminosityBlock&, edm::EventSetup const&) const final;
+  std::shared_ptr<testinter::LumiCache> globalBeginLuminosityBlock(edm::LuminosityBlock const&,
+                                                                   edm::EventSetup const&) const final;
+  void streamBeginLuminosityBlock(edm::StreamID, edm::LuminosityBlock const&, edm::EventSetup const&) const final;
+  void streamEndLuminosityBlock(edm::StreamID, edm::LuminosityBlock const&, edm::EventSetup const&) const final;
+  void globalEndLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&) const final {}
+  void globalEndLuminosityBlockProduce(edm::LuminosityBlock&, edm::EventSetup const&) const final;
+
+private:
+  edm::EDPutTokenT<edmtest::ThingCollection> const token_;
+  edm::EDPutTokenT<edmtest::ThingCollection> const brToken_;
+  edm::EDPutTokenT<edmtest::ThingCollection> const erToken_;
+  edm::EDPutTokenT<edmtest::ThingCollection> const blToken_;
+  edm::EDPutTokenT<edmtest::ThingCollection> const elToken_;
+
+  std::string config_;
+
+  mutable testinter::StreamCache* stream0Cache_ = nullptr;
+  mutable std::atomic<testinter::StreamCache*> availableForBeginLumi_;
+  mutable std::atomic<unsigned int> lastLumiIndex_ = 0;
+};
+
+TestInterProcessProd::TestInterProcessProd(edm::ParameterSet const& iPSet)
+    : token_{produces<edmtest::ThingCollection>()},
+      brToken_{produces<edmtest::ThingCollection, edm::Transition::BeginRun>("beginRun")},
+      erToken_{produces<edmtest::ThingCollection, edm::Transition::EndRun>("endRun")},
+      blToken_{produces<edmtest::ThingCollection, edm::Transition::BeginLuminosityBlock>("beginLumi")},
+      elToken_{produces<edmtest::ThingCollection, edm::Transition::EndLuminosityBlock>("endLumi")},
+      config_{iPSet.getUntrackedParameter<std::string>("python_config")} {}
+
+std::unique_ptr<testinter::StreamCache> TestInterProcessProd::beginStream(edm::StreamID iID) const {
+  auto const label = moduleDescription().moduleLabel();
+
+  using namespace std::string_literals;
+
+  std::string config = R"_(from FWCore.TestProcessor.TestProcess import *
+process = TestProcess()
+)_";
+  config += "process."s + label + "=" + config_ + "\n";
+  config += "process.moduleToTest(process."s + label + ")\n";
+  config += R"_(
+process.add_(cms.Service("InitRootHandlers", UnloadRootSigHandler=cms.untracked.bool(True)))
+  )_";
+
+  auto cache = std::make_unique<testinter::StreamCache>(config, iID.value());
+  if (iID.value() == 0) {
+    stream0Cache_ = cache.get();
+
+    availableForBeginLumi_ = stream0Cache_;
```

Comment:
I think it would be good to explain somewhere the mechanism with `availableForBeginLumi_` and `lastLumiIndex_`.

Reply:
It is now with the member variable declaration.